src/TableData.cpp

```diff
@@ -805,3 +809,169 @@ K parseArrowData(K char_array, K options)
+#else
+  if (!kx::arrowkdb::IsKdbString(orc_file))
+    return krr((S)"orc_file not 11h or 0 of 10h");
+
+  // Parse the options
+  auto read_options = kx::arrowkdb::KdbOptions(options, kx::arrowkdb::Options::string_options, kx::arrowkdb::Options::int_options);
+
+  // Use memmap
+  int64_t use_mmap = 0;
+  read_options.GetIntOption(kx::arrowkdb::Options::USE_MMAP, use_mmap);
+
+  // Type mapping overrides
+  kx::arrowkdb::TypeMappingOverride type_overrides{ read_options };
+
+  std::shared_ptr<arrow::io::RandomAccessFile> infile;
+  if (use_mmap) {
+    PARQUET_ASSIGN_OR_THROW(
+      infile,
+      arrow::io::MemoryMappedFile::Open(kx::arrowkdb::GetKdbString(orc_file),
+        arrow::io::FileMode::READ));
+  } else {
+    PARQUET_ASSIGN_OR_THROW(
+      infile,
+      arrow::io::ReadableFile::Open(kx::arrowkdb::GetKdbString(orc_file),
+        arrow::default_memory_pool()));
+  }
+  // Open ORC file reader
+  auto maybe_reader = arrow::adapters::orc::ORCFileReader::Open(infile, arrow::default_memory_pool());
+
+  std::unique_ptr<arrow::adapters::orc::ORCFileReader> reader = std::move(maybe_reader.ValueOrDie());
+
+  // Read entire file as a single Arrow table
+  auto maybe_table = reader->Read();
+
+  std::shared_ptr<arrow::Table> table = maybe_table.ValueOrDie();
+
+  const auto schema = table->schema();
+  SchemaContainsNullable(schema);
+  const auto col_num = schema->num_fields();
+  K data = ktn(0, col_num);
+  for (auto i = 0; i < col_num; ++i) {
+    auto chunked_array = table->column(i);
+    kK(data)[i] = kx::arrowkdb::ReadChunkedArray(chunked_array, type_overrides);
+  }
+
+  return data;
+#endif
+
+  KDB_EXCEPTION_CATCH;
+}
+
+K readORCSchema(K orc_file)
+{
+  KDB_EXCEPTION_TRY;
+
+#ifdef _WIN32
+  return krr((S)"ORC files are not supported on Windows");
+#else
+  if (!kx::arrowkdb::IsKdbString(orc_file))
+    return krr((S)"orc_file not 11h or 0 of 10h");
+
+  std::shared_ptr<arrow::io::ReadableFile> infile;
+  PARQUET_ASSIGN_OR_THROW(
+    infile,
+    arrow::io::ReadableFile::Open(kx::arrowkdb::GetKdbString(orc_file),
+      arrow::default_memory_pool()));
+
+  auto maybe_reader = arrow::adapters::orc::ORCFileReader::Open(infile, arrow::default_memory_pool());
+
+  std::unique_ptr<arrow::adapters::orc::ORCFileReader> reader = std::move(maybe_reader.ValueOrDie());
+
+  auto maybe_schema = reader->ReadSchema();
+
+  std::shared_ptr<arrow::Schema> schema = maybe_schema.ValueOrDie();
+  // Add each field from the table to the field store
+  // Add each datatype from the table to the datatype store
+  //const auto schema = table->schema();
+  SchemaContainsNullable(schema);
+  for (auto field : schema->fields()) {
+    kx::arrowkdb::GetFieldStore()->Add(field);
+    kx::arrowkdb::GetDatatypeStore()->Add(field->type());
+  }
+
+  // Return the new schema_id
+  return ki(kx::arrowkdb::GetSchemaStore()->Add(schema));
+#endif
+
+  KDB_EXCEPTION_CATCH;
+}
+
+
+K writeORC(K orc_file, K schema_id, K array_data, K options)

```

Comment:
Can we support null mapping and null bitmap?
I think so
Model off the new writeParquet / writeArrow / serializeArrow

Reply:
Will come in separate code-review
https://kxl.atlassian.net/browse/KXI-24287